### PR TITLE
Validate product variant and its product before saving them (#148)

### PR DESCRIPTION
**Pull request by mmenozzi**
Fixes: #148 #137 (maybe others)
Replaces: #149 

Here we validate the product variant and its product before saving them. This way every validation error that would be thrown in the Sylius backend is triggered during the importer before reaching Doctrine to avoid subsequent "EntityManager is closed" errors.

### CODE REVIEW

**Comment by mmenozzi** [comment]
Errors are in the Product Importer integration tests because now we validates the product and does not allow to have product invalid state.

File: tests/Behat/Context/Cli/ImportCommandContext.php [hunk 1]
```diff
@@ -167,6 +181,16 @@ public function iShouldBeNotifiedThatTheImporterDoesNotExists(): void
         Assert::regex($throwable->getMessage(), '/Importer ".*?" does not exists/');
     }
 
+    /**
+     * @Then /^I should get an error about product name cannot be empty$/
+     */
+    public function iShouldNotGetAnyError(): void
```
Comment (LucaGallinari) [inline tests/Behat/Context/Cli/ImportCommandContext.php:187]:
```suggestion
    public function iShouldGetAnErrorAboutProductNameNotEmpty(): void
```

**Review by LucaGallinari — APPROVED** [review]
Ho segnato una cosetta

File: config/services.xml [hunk 2]
```diff
@@ -143,5 +144,11 @@
             <tag name="messenger.message_handler"/>
         </service>
 
+        <service id="webgriffe_sylius_akeneo.product.validator" class="Webgriffe\SyliusAkeneoPlugin\Product\Validator">
```
Comment (lruozzi9) [inline config/services.xml:147]: Maybe a more right name could be `webgriffe_sylius_akeneo.validator.product`? And as a consequence also the namespace should become `Webgriffe\SyliusAkeneoPlugin\Validator\Product`?

File: features/importing_products.feature [hunk 3]
```diff
@@ -4,13 +4,11 @@ Feature: Importing products
...
-  Background:
-    Given there is a product "BRAIDED_HAT_M" on Akeneo
-    And there is a product "BRAIDED_HAT_L" on Akeneo
-
   @cli
   Scenario: Importing single product model and its variants
-    Given the store operates on a single channel
+    Given there is a product "BRAIDED_HAT_M" on Akeneo
+    And there is a product "BRAIDED_HAT_L" on Akeneo
+    And the store operates on a single channel
```
Comment (lruozzi9) [inline features/importing_products.feature:11]: To me, this step could be moved to the background, am I right?

File: src/Product/ValidationException.php [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Webgriffe\SyliusAkeneoPlugin\Product;
```
Comment (lruozzi9) [inline src/Product/ValidationException.php:5]:
```suggestion
namespace Webgriffe\SyliusAkeneoPlugin\Product\Exception;
```

File: src/ValueHandler/ImmutableSlugValueHandler.php [hunk 5]
```diff
@@ -60,7 +60,9 @@ public function handle($subject, string $attribute, array $value): void
         foreach ($value as $item) {
             $localeCode = $item['locale'];
             $valueToSlugify = $item['data'];
-            Assert::string($valueToSlugify);
+            if (!is_string($valueToSlugify)) {
+                continue;
+            }
```
Comment (lruozzi9) [inline src/ValueHandler/ImmutableSlugValueHandler.php:65]: I'm not sure about this, maybe throwing an exception is a more secure way to handle this case? It sounds like an error if a value is not a string...

**Review by lruozzi9 — CHANGES_REQUESTED** [review]
Very good!! 🚀🎸

File: config/services.xml (hunk 2 above)
Comment (mmenozzi) [reply config/services.xml:147]: Mmmm no... All Product importer stuff are in the `Webgriffe\SyliusAkeneoPlugin\Product` namespace.

**Review by lruozzi9 — APPROVED** [review]
Ok!

**Pull request merged**